Task 2

zavijavasoft opened the pull request.

Commit:
Подготовка .gradle и Proguard к релизу

Commit:
Добавлена поддержка RxJava и RxAndroid

Commit:
В модель добавлено хранилище вычисленного баланса

Commit:
Изменен интерфейс FinanceTracker и его основная реализация

Добавлено сохранение вычисленного баланса в отдельной репе
Баланс теперь сначала берется из хранилища, а потом вычисляется
Тест добавлен больше для проверки правильности моих представлений о том,
как работает Rx

Commit:
В модель добавлены сущности "Статья" и "Счет"

Commit:
Добавлен новый фрагмент, отражающий операции по счетам

Добавлен сам фрагмент, презентер, добавлены операции по вычислению
остатков на счетах

Commit:
Исправлен баг с неправильным учетом отрицательных сумм

Commit:
Коррекция дизайна фрагмента операций

Commit:
Добавлен механизм Drag'n'Drop

Commit:
Переразмещение файлов исходников для удобства

Commit:
Добавлено определение типа требуемой операции

Commit:
Перевод на Rx2. Переделан конвертер валют

Commit:
Добавлено скачивание курса валют с сайта центробанка

Добавлена соответствующая Retrofit + XML - реализация монитора валют

Commit:
Добавлены экспериментальные фишки котлина для халявного @Parcelize

ну, и для корутин, если понадобятся

Commit:
Добавлена черновая активити для ввода сумм

Доработан механизм подтверждения транзакции

Commit:
Добавлен фрагмент с отображением транзакций (пока черновой вариант)

Commit:
Изменен начальный фрагмент (добавлен расчет баланса во всех валютах)

Исправлен баг с округлением курсов при загрузке из интернета

Commit:
Добавлен игрушечный круговой чарт

Commit:
Фиксация глюка гита в AS

Commit:
Фиксация удаленного файла

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/BalancePresenterImpl.kt:
```diff
@@ -13,14 +16,26 @@ class BalancePresenterImpl @Inject constructor(private val tracker: FinanceTrack
+    override fun needUpdate() {
+        val currencyRatios = currencyMonitor.pull()
+        val transactions = tracker.retrieveTransactions()
+        Single.zip(currencyRatios, transactions,
```
У меня есть такая привычка: всегда сохранять то, что возвращается из функции subscribe и потом в зависимости от жизненного цикла отменять subscribe.
Тут у тебя возвращается кажется Disposable и его можно хранить в CompositeDisposable, а в onStop (onDestory, onDetach или что-нибудь такое) отменять подписки.

zavijavasoft replied:
Все замечания по Rx учту и буду применять

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/BalancePresenterImpl.kt:
```diff
@@ -13,14 +16,26 @@ class BalancePresenterImpl @Inject constructor(private val tracker: FinanceTrack
+                .subscribe { it ->
+                    it.subscribe {
+                        val list = mutableListOf<BalanceChunk>()
+                        for (currency in it.balance.keys) {
```
тут можно заменить на map

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/di/DIModules.kt:
```diff
@@ -59,3 +67,32 @@ class TransactionsStorageModule {
 }
 
+@Module
+class BalanceStorageModule {
```
А в чем преимущество именно такого разбиения по модулям? Просто интересно)

zavijavasoft replied:
Честно, не знаю, как их по другому разбивать. Даггер мне не дает биндить "листовые" зависимости, их приходится провайдить. К тому же тут все синглтоны, поэтому модель плоская

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/OperationPresenterImpl.kt:
```diff
@@ -0,0 +1,140 @@
+    override fun requireIncomeTransaction(articleId: Long, accountId: Long) {
+
+        tracker.getAccountsList().subscribe { it ->
+            val account = it.find { it.id == accountId }
```
Вроде бы это типичная задача: получить элемент по его ID. Может прямо у трэкера сделать сооветствующий метод?

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/core/OperationPresenterImpl.kt:
```diff
@@ -0,0 +1,140 @@
+                BiFunction { accountsList: List<AccountEntity>,
+                             restMap: Map<Long, Float> ->
+                    val account = accountsList.find { it.id == accountId }
+                    if (account != null) {
```
А если null, то что должно произойти?
Тут либо null не может быть либо выполнение завершится не успешно. Но ты не узнаешь об этом, т.к. блок else пустой.

zavijavasoft replied:
Это попытка смарт-каста AccountEntity?, который возвращает find(). Теоретически тут не может быть null

HotIceCream replied:
Тогда лучше выбросить исключение в else. Либо account!! написать. 
Иначе в какой-то момент при разработке у тебя это условие может нарушиться и ты не будешь понимать почему у тебя не работает приложение. А с исключением - ты сразу заметишь в каком месте ошибка.

Если подобное поведение произойдет у пользователя - то тоже лучше пусть приложение упадет и ты увидишь отчет в метрике (и поправишь проблему), чем если приложение не будет работать, но ты об этом не будешь знать.

zavijavasoft replied:
Принято, спасибо

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/di/DIModules.kt:
```diff
@@ -13,12 +15,18 @@ import javax.inject.Singleton
+
+
 typealias ConcreteTransactionStorageImpl = StubTransactionStorageImpl
-typealias ConcreteCurrencyMonitorImpl = StubCurrencyMonitorImpl
```
не знал о таком подходе, интересный 👍

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/BalanceModel.kt:
```diff
@@ -0,0 +1,48 @@
+
+fun BalanceEntity.toCurrenciesString(): String {
+    val sb = StringBuffer()
+    for (currency in balance.keys) {
```
а тут не подойдет какой-нибудь join? Например Array.joinTo или в каких-нибудь TextUtils.join
Возможно в котлине еще что-то есть.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/FinanceTracker.kt:
```diff
@@ -1,15 +1,21 @@
+import io.reactivex.Flowable
+import io.reactivex.Single
+
 interface FinanceTracker {
```
У этого интерфейса по-моему очень много обязанностей. Возможно имеет смысл разбить его на несколько разных классов.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/FinanceTrackerImpl.kt:
```diff
@@ -15,50 +21,80 @@ class FinanceTrackerImpl @Inject constructor(var storage: TransactionStorage) :
+
+
+        val initialBalance = balanceStorage.getBalance()
+        val calculatedBalance: Single<BalanceEntity> = Single.fromCallable<BalanceEntity> {
```
Я бы это делал по-другому, например так:
`
Single.fromCallable() { calculateAll() }
    .flatMap { balanceStorage.setBalance(newBalance).toSingle(it) }
`
К сожалению, в последнее время не работаю с rx, поэтому мог ошибиться в синтаксисе,  но надеюсь идея будет понятна.

zavijavasoft replied:
Вот здесь спасибо большое. Я не успел нагуглить такое решение)

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/FinanceTrackerImpl.kt:
```diff
@@ -68,8 +104,48 @@ class FinanceTrackerImpl @Inject constructor(var storage: TransactionStorage) :
-    override fun listCurrenciesInAccounts(): List<String> {
-        return transactions.asSequence().map { it.currency }.distinct().toList()
+    fun updateArticlesAndAccounts() {
+        val accountlist = accountsStorage.getAccounts().blockingGet()
```
blocking* функции если у тебя используются в коде (за исключением тестов), то значит что-то ты не так делаешь с rx. В идеале у тебя везде должны быть Observable\Single\Flowable и ни где не быть blockinGet()

zavijavasoft replied:
Это от недостатка знаний о связывании Rx в цепочки... Но я учусь

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/SharedPrefBalanceStorageImpl.kt:
```diff
@@ -0,0 +1,51 @@
+    lateinit var context: Context
+
+    init {
+        YaFinaApplication.component.inject(this)
```
Вместо этого нужно использовать инжекцию через конструктор.

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/model/SharedPrefBalanceStorageImpl.kt:
```diff
@@ -0,0 +1,51 @@
+
+    override fun getBalance(): Single<BalanceEntity> {
+        return Single.create<BalanceEntity> {
+            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
```
почему нельзя сделать это полемкласса?

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/ui/OpenTransactionActivity.kt:
```diff
@@ -0,0 +1,99 @@
+
+    override fun onDestroy() {
+        super.onDestroy()
+        unbinder.unbind()
```
в activity unbinder особо смысла не имеет. Он больше нужен во фрагментах.

zavijavasoft replied:
для симметрии разве что. Во фрагментах везде есть)

HotIceCream replied:
Ок, самое главное что бы ты понимал почему оно нужно во фрагментах:)

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/ui/YaFinaActivity.kt:
```diff
@@ -80,6 +106,32 @@ class YaFinaActivity : MvpAppCompatActivity(), MainView {
             }
         }
 
+        buttonTransactionsList.setOnClickListener {
```
лучше сразу удалять закомментированый код. И уж точно не оставлять его в PR

HotIceCream commented on app/src/main/java/com/zavijavasoft/yafina/utils/Utils.kt:
```diff
@@ -1,5 +1,44 @@
+object ColorSelector {
+
+    val mColors = arrayOf("#39add1", // light blue
+            "#3079ab", // dark blue
```
лучше в ресурсах хранить цвета. В том числе в виде массива

HotIceCream commented on app/src/test/java/com/zavijavasoft/yafina/FinanceTrackerImplTest.kt:
```diff
@@ -2,37 +2,62 @@ package com.zavijavasoft.yafina
+        val logger = Logger.getAnonymousLogger()
+    }
+
+    val currencyStorage = StubCurrencyStorageImpl()
```
К примеру, если попробовать разбить FinanceTracker на несколько классов - у него не будет столько зависимостей, а значит будет проще тестировать.

Stub реализации это хорошо. Но обычно можно обойтись mockito

zavijavasoft replied:
Эти стабы для "хардкодных" частей приложения. Они не только в тестах используются, но и в каркасе. 
Идея использования для этой цели мокито мне в голову не приходила)

HotIceCream replied:
В тестах для этой цели мокито отлично заходит. В приложении понятное дело, что нужны стабы

HotIceCream reviewed (commented):
Думаю стоит получитьше изучить api rxjava, там много разных функций, которые могут пригодиться.

Commit:
Поправлен критический баг

Баг: Moxy пыталась повторить requireTransaction после смены
конфигурации, воскрешая убитую активити,но без корректного интента, что
при повторной смене конфигурации приводило к крашу
В дальнейшем я хотел превратить эту активити во фрагмент, где эта
проблема была бы неактуальной
Решено сменой стратегии для данного метода

Commit:
исправлены некоторые замечания CR

Инжекция контекста в SharedPrefBalanceStorageImpl сделана через
конструктор
Удален закомменченный код
